clang/test/CIR/CodeGen/no-prototype.c

```diff
@@ -0,0 +1,49 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+// No-proto definition followed by a correct call.
+int noProto0(x) int x; { return x; }
+int test0(int x) {
+  // CHECK: cir.func @test0
+  return noProto0(x); // We know the definition. Should be a direct call.
+  // CHECK: %{{.+}} = cir.call @noProto0(%{{.+}})
+}
+
+// Declaration without prototype followed by its definition, then a correct call.
+//
+// Call to no-proto is made after definition, so a direct call can be used.
```

Comment:
Where's `#cir.no_proto` being used / checked for?

Reply:
I've added `#cir.no_proto` on the following commit.

This commit handles only valid calls, which do not need to bypass CIR's type checking.

The following commit deals with bypassing type-checking on invalid calls.

Reply:
I thought you needed the attribute to decide if you want to materialize a new version now that definition appeared, but are you doing that by looking at the Decl directly, is my understanding right?

Reply:
> doing that by looking at the Decl directly, is my understanding right?

That's right.

Once a function definition with the same name appears, uses of the no-proto one are replaced. 

That whole mess of getting a func pointer, doing a bitcast, and then an indirect call was actually only required to make the types match after the function declaration was replaced. But if the call is correct, it can be called directly, only when it is a wrong call that the bitcast was required to bypass type checking.

With the `#cir.no_proto` attribute, we do this bypass by just checking if the function has the attribute.

Reply:
> With the #cir.no_proto attribute, we do this bypass by just checking if the function has the attribute.

But do all no-proto FuncOps marked with that attribute when first found? For consistency they should, even if they end up being erased.

Reply:
They do, except for one case.

When a no-proto declaration is found, we add the tag. When we see the definition and overwrite the no-proto declaration, the attribute is carried over to the new function.

The notable exception is when a no-proto reaches a definition with no call in between. In this case, invalid calls actually break Clang.

This, for example, does not compile:
```c
int noProto();
int noProto(int x) { return 0; }
int main(void) {
  return noProto(/*missing arg*/);
}
```
 
 However, this does:
```c
int noProto();
int main(void) {
  return noProto(/*missing arg*/);
}
int noProto(int x) { return 0; }
```

Reply:
> The notable exception is when a no-proto reaches a definition with no call in between. In this case, invalid calls actually break Clang.

Did you break as in crash or is the code rejected? Seems like its reject https://godbolt.org/z/e9zabM49b
So it's probably not interesting for codegen?

Reply:
Sorry. I meant "code reject".

> So it's probably not interesting for codegen?

Oh, there are no tests like the one above. But there is one similar where the call is valid. The point of this comparison it to explain why there is one test where the `cir.no_proto` flag is not set (despite having a no-proto decl). See here: https://github.com/llvm/clangir/pull/130/commits/e014e1aa142b30f91c0b3566d6763bfb69b4f1f6.

You asked if all no-proto ops are marked with the attribute, right? They are, but there is this caveat :point_up:.